server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java

```diff
@@ -0,0 +1,314 @@
+  private static final String HOST_BALANCER_POOL_RECHECK_DEFAULT = "5m";
+  public static final String HOST_BALANCER_REGEX_USING_IPS = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.is.ip";
+  protected static final String DEFAULT_POOL = "HostTableLoadBalancer.ALL";
+
+  protected long oobCheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_OOB_DEFAULT);
+  protected long poolRecheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_POOL_RECHECK_DEFAULT);
+
+  private Map<String,String> tableIdToTableName = null;
+  private Map<String,Pattern> poolNameToRegexPattern = null;
+  private long lastOOBCheck = System.currentTimeMillis();
+  private long lastPoolRecheck = 0;
+  private boolean isIpBasedRegex = false;
+  private Map<String,SortedMap<TServerInstance,TabletServerStatus>> pools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+
+  /**
+   * Group the set of current tservers by pool name. Tservers that don't match a regex are put into a default ppol.
+   *
+   * @param current
+   *          map of current tservers
+   * @return current servers grouped by pool name, if not a match it is put into a default pool.
+   */
+  protected synchronized Map<String,SortedMap<TServerInstance,TabletServerStatus>> splitCurrentByRegex(SortedMap<TServerInstance,TabletServerStatus> current) {
+    if ((System.currentTimeMillis() - lastPoolRecheck) > poolRecheckMillis) {
+      Map<String,SortedMap<TServerInstance,TabletServerStatus>> newPools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+      for (Entry<TServerInstance,TabletServerStatus> e : current.entrySet()) {
+        String tableName = getPoolNameForHost(e.getKey().host());
+        if (!newPools.containsKey(tableName)) {
+          newPools.put(tableName, new TreeMap<TServerInstance,TabletServerStatus>());
+        }
+        newPools.get(tableName).put(e.getKey(), e.getValue());
+      }
+      // Ensure that no host is in more than one pool
+      // TODO: I'm not sure that I need to check for disjoint as the call to getPoolNameForHost checks for more than one match
+      boolean error = false;
+      for (SortedMap<TServerInstance,TabletServerStatus> s1 : newPools.values()) {
+        for (SortedMap<TServerInstance,TabletServerStatus> s2 : newPools.values()) {
+          if (s1 == s2) {
+            continue;
+          }
+          if (!Collections.disjoint(s1.keySet(), s2.keySet())) {
+            LOG.error("Pools are not disjoint: {}, there is a problem with your regular expressions. Putting all servers in default pool", newPools);
+            error = true;
+          }
+        }
+      }
+      if (error) {
+        // Put all servers into the default pool
+        newPools.clear();
+        newPools.put(DEFAULT_POOL, new TreeMap<TServerInstance,TabletServerStatus>());
+        for (Entry<TServerInstance,TabletServerStatus> e : current.entrySet()) {
```

Comment:
Nitpick, so ignore if you want, but you could replace this loop with newPools.putAll(current). (Same comment as above about the comparator also applies.)